src/qutip_cupy/dense_functions.py

```diff
@@ -29,3 +29,33 @@ def trace_cupydense(cp_arr):
     # @TODO: whnen qutip allows it we should remove this call to item()
     # as it takes a time penalty commmunicating data from GPU to CPU.
     return cp.trace(cp_arr._cp).item()
+
+
+hermcheck_kernel = cp.RawKernel(
+    r"""
+    #include <cupy/complex.cuh>
+    extern "C" __global__
+    void hermcheck(const complex<double>* x1,const int size,
+                    const double tol, unsigned int* y) {
+        int tidx = blockDim.x * blockIdx.x + threadIdx.x;
+        int tidy = blockDim.y * blockIdx.y + threadIdx.y;
+        if((tidx < size) & (tidy < size)){
+           atomicAnd(y,int(norm(x1[tidx*size+tidy] - conj(x1[tidy*size+tidx])) < tol));
+        };
```

Comment:
1. nitpick: the IDs are naturally `unsigned int`.
2. We need logical AND (`&&`), not bitwise AND (`&`).
3. What happens if `size` is excessively large (`> gridDim.x*blockDim.x`)?
```suggestion
        for (unsigned int tidx = blockDim.x * blockIdx.x + threadIdx.x; tidx < size; tidx += gridDim.x * blockDim.x) {
          for (unsigned int tidy = blockDim.y * blockIdx.y + threadIdx.y; tidy < size; tidy += gridDim.y * blockDim.y) {
            atomicAnd(y, int(norm(x1[tidx*size+tidy] - conj(x1[tidy*size+tidx])) < tol));
          }
        };
```

Reply:
One more problem with the previous `atomicAdd` version was that we used the `int` version, which cannot handle if `size` > 2^31 -1. Even with `unsigned int` the cap is 2^32 - 1. To be rock solid the kernel had to be templated with `y`'s type, so that the `unsigned long long` version can be used for `size>=2^32`. Unfortunately, this means much slower performance (wider bit width & further serialization).

Reply:
I added this suggestions to the new reduction version. Thank you Leo :)

Reply:
btw it's called ["grid-stride loop"](https://developer.nvidia.com/blog/cuda-pro-tip-write-flexible-kernels-grid-stride-loops/).